diff --git a/epan/dissectors/packet-ssl-utils.c b/epan/dissectors/packet-ssl-utils.c
index 8f85f11..a58ccfb 100644
--- a/epan/dissectors/packet-ssl-utils.c
+++ b/epan/dissectors/packet-ssl-utils.c
@@ -1034,6 +1034,7 @@ const value_string tls_hello_extension_types[] = {
     { 13, "signature_algorithms" },  /* RFC 5246 */
     { 14, "use_srtp" },
     { SSL_HND_HELLO_EXT_HEARTBEAT, "Heartbeat" },  /* RFC 6520 */
+    { SSL_HND_HELLO_EXT_ENCRYPT_THEN_MAC_TYPE, "Encrypt then MAC" }, /* RFC 7366 */
     { SSL_HND_HELLO_EXT_EXTENDED_MASTER_SECRET_TYPE, "Extended Master Secret" }, /* https://tools.ietf.org/html/draft-ietf-tls-session-hash-01 */
     { 35, "SessionTicket TLS" },  /* RFC 4507 */
     { SSL_HND_HELLO_EXT_NPN, "next_protocol_negotiation"}, /* http://technotes.googlecode.com/git/nextprotoneg.html */
diff --git a/epan/dissectors/packet-ssl-utils.h b/epan/dissectors/packet-ssl-utils.h
index 5968b8e..82dc39b 100644
--- a/epan/dissectors/packet-ssl-utils.h
+++ b/epan/dissectors/packet-ssl-utils.h
@@ -153,6 +153,7 @@
 #define SSL_HND_HELLO_EXT_EC_POINT_FORMATS            0x000b
 #define SSL_HND_HELLO_EXT_SIG_HASH_ALGS               0x000d
 #define SSL_HND_HELLO_EXT_HEARTBEAT                   0x000f
+#define SSL_HND_HELLO_EXT_ENCRYPT_THEN_MAC_TYPE       0x0016
 #define SSL_HND_HELLO_EXT_EXTENDED_MASTER_SECRET_TYPE 0x0017
 #define SSL_HND_HELLO_EXT_RENEG_INFO                  0xff01
 #define SSL_HND_HELLO_EXT_NPN                         0x3374
