diff --git a/ui/gtk/main.c b/ui/gtk/main.c
index 2621317..afa7cf3 100644
--- a/ui/gtk/main.c
+++ b/ui/gtk/main.c
@@ -2069,50 +2069,6 @@ read_configuration_files(char **gdp_path, char **dp_path)
   return prefs_p;
 }
 
-/*  Check if there's something important to tell the user during startup.
- *  We want to do this *after* showing the main window so that any windows
- *  we pop up will be above the main window.
- */
-static void
-#ifdef _WIN32
-check_and_warn_user_startup(gchar *cf_name)
-#else
-check_and_warn_user_startup(gchar *cf_name _U_)
-#endif
-{
-  gchar               *cur_user, *cur_group;
-  gpointer             priv_warning_dialog;
-
-  /* Tell the user not to run as root. */
-  if (running_with_special_privs() && recent.privs_warn_if_elevated) {
-    cur_user = get_cur_username();
-    cur_group = get_cur_groupname();
-    priv_warning_dialog = simple_dialog(ESD_TYPE_WARN, ESD_BTN_OK,
-      "Running as user \"%s\" and group \"%s\".\n"
-      "This could be dangerous.\n\n"
-      "If you're running Wireshark this way in order to perform live capture, "
-      "you may want to be aware that there is a better way documented at\n"
-      "http://wiki.wireshark.org/CaptureSetup/CapturePrivileges", cur_user, cur_group);
-    g_free(cur_user);
-    g_free(cur_group);
-    simple_dialog_check_set(priv_warning_dialog, "Don't show this message again.");
-    simple_dialog_set_cb(priv_warning_dialog, priv_warning_dialog_cb, NULL);
-  }
-
-#ifdef _WIN32
-  /* Warn the user if npf.sys isn't loaded. */
-  if (!stdin_capture && !cf_name && !npf_sys_is_running() && recent.privs_warn_if_no_npf && get_os_major_version() >= 6) {
-    priv_warning_dialog = simple_dialog(ESD_TYPE_WARN, ESD_BTN_OK,
-      "The NPF driver isn't running.  You may have trouble\n"
-      "capturing or listing interfaces.");
-    simple_dialog_check_set(priv_warning_dialog, "Don't show this message again.");
-    simple_dialog_set_cb(priv_warning_dialog, npf_warning_dialog_cb, NULL);
-  }
-#endif
-
-}
-
-
 /* And now our feature presentation... [ fade to music ] */
 int
 main(int argc, char *argv[])
@@ -3022,7 +2978,6 @@ main(int argc, char *argv[])
      up on top of us. */
   if (cf_name) {
     show_main_window(TRUE);
-    check_and_warn_user_startup(cf_name);
     if (rfilter != NULL) {
       if (!dfilter_compile(rfilter, &rfcode)) {
         bad_dfilter_alert_box(top_level, rfilter);
@@ -3100,7 +3055,6 @@ main(int argc, char *argv[])
           dfilter_free(rfcode);
         cfile.rfcode = NULL;
         show_main_window(FALSE);
-        /* Don't call check_and_warn_user_startup(): we did it above */
         set_menus_for_capture_in_progress(FALSE);
         set_capture_if_dialog_for_capture_in_progress(FALSE);
       }
@@ -3117,7 +3071,6 @@ main(int argc, char *argv[])
       }
       /* "-k" was specified; start a capture. */
       show_main_window(FALSE);
-      check_and_warn_user_startup(cf_name);
 
       /* If no user interfaces were specified on the command line,
          copy the list of selected interfaces to the set of interfaces
@@ -3135,7 +3088,6 @@ main(int argc, char *argv[])
       }
     } else {
       show_main_window(FALSE);
-      check_and_warn_user_startup(cf_name);
       set_menus_for_capture_in_progress(FALSE);
       set_capture_if_dialog_for_capture_in_progress(FALSE);
     }
@@ -3146,7 +3098,6 @@ main(int argc, char *argv[])
     }
 #else /* HAVE_LIBPCAP */
     show_main_window(FALSE);
-    check_and_warn_user_startup(cf_name);
     set_menus_for_capture_in_progress(FALSE);
     set_capture_if_dialog_for_capture_in_progress(FALSE);
 #endif /* HAVE_LIBPCAP */
