diff --git a/tshark.c b/tshark.c
index 8c8de07..8cc4fae 100644
--- a/tshark.c
+++ b/tshark.c
@@ -286,7 +286,7 @@ print_usage(gboolean print_ver)
 #endif
   /*fprintf(output, "\n");*/
   fprintf(output, "Input file:\n");
-  fprintf(output, "  -r <infile>              set the filename to read from (no pipes or stdin!)\n");
+  fprintf(output, "  -r <infile>              set the filename to read from (no stdin!)\n");
 
   fprintf(output, "\n");
   fprintf(output, "Processing:\n");
diff --git a/wiretap/file_wrappers.c b/wiretap/file_wrappers.c
index c670c6e..e359cd7 100644
--- a/wiretap/file_wrappers.c
+++ b/wiretap/file_wrappers.c
@@ -1046,7 +1046,8 @@ file_seek(FILE_T file, gint64 offset, int whence, int *err)
 
 	/* if within raw area while reading, just go there */
 	if (file->compression == UNCOMPRESSED && file->pos + offset >= file->raw
-			&& (offset < 0 || offset >= file->have) /* seek only when we don't have that offset in buffer */)
+			&& (offset < 0 || offset >= file->have) /* seek only when we don't have that offset in buffer */
+			&& (file->fast_seek) /* seek only when random access is supported */)
 	{
 		if (ws_lseek64(file->fd, offset - file->have, SEEK_CUR) == -1) {
 			*err = errno;
